### DISCUSSION TITLE
[idea]add %debug or %gdlv(solved with 2 easy solutions)

### DISCUSSION THREAD

**Discussion by HaveF** [janpfeifer/gonb#122] [Ideas]
[go-delve/delve](https://github.com/go-delve/delve) is the de facto debug tool for Go.

[Debugging Go Code with GDB - The Go Programming Language](https://go.dev/doc/gdb)
> Note that [Delve](https://github.com/go-delve/delve) is a better alternative to GDB when debugging Go programs built with the standard toolchain. It understands the Go runtime, data structures, and expressions better than GDB. Delve currently supports Linux, OSX, and Windows on amd64. For the most up-to-date list of supported platforms, please see [the Delve documentation](https://github.com/go-delve/delve/tree/master/Documentation/installation).

I find it useful. It does not have gui, but it has [gui plugin](https://github.com/go-delve/delve/blob/master/Documentation/EditorIntegration.md#alternative-uis). After a quick dive, I find [aarzilli/gdlv](https://github.com/aarzilli/gdlv) has more stars and editor-independent.

For now, when I want use it, I have to check where the generated main.go is by `!*pwd`. Then I enter the command in the terminal:

```
gdlv debug /var/folders/wd/xsn02ct128g0zv9_j1yj0wdc0000gn/T/gonb_a79acdb0/main.go
```

I found this process boring. Could it possible support it as a special command in gonb, like `%debug` or `%gdlv`, something like that.

Someone prefer `fmt.Println`, I believe it is just because debugger is hard to use...When we can use debugger like `gdlv` freely, I believe it can save a lot time and typing(`fmt.Println`)

**Comment by janpfeifer** [comment]
Very interesting.

I haven't yet studied the debug protocol for Jupyter, but it's definitely a nice to have feature.

There is an issue opened about this in issue #97 , maybe add the pointers there ?

**Reply by HaveF** [reply]
@janpfeifer Maybe I didn't make it clear. There is no need to study the jupyter debug protocol.
Just pass the current path to gdlv and start it
For example, in fact `%gdlv` is actually a shortcut of

`!gdlv debug {*pwd}/main.go`

**Just debug directly in a separate gdlv**. The problem now is that I have no way to get the result of `!*pwd`. `{*pwd}` is an invalid expression.

@Michael-F-Ellis Hi, What do you think of this debug solution? Is it simple?

**Reply by janpfeifer** [reply]
Apologies, then I think I'm not following. 

I haven't used Delve -- probably I should have duh ... but I wonder if it handles well Go programs with CGO (what I've been doing a lot recently with GoMLX). 

But thinking of `gdb`: it requires an interactive shell (with a responsive `stdin`/`stdout`). In Jupyter notebooks one doesn't get that. Only the `stdout` is responsive. There is a not-very-convenient way to get a `stdin` input (for passwords), but it requires the user to click on a text box and type ... It wouldn't be interactive like a terminal.

Just do `!cat` in a cell, and compare it with `cat` in a terminal.

Now ... it would be possible to create an interactive terminal as a widget within the notebook. But that is a project in itself.

Or are you suggesting something different and I'm missing it ? (apologies)

**Comment by HaveF** [comment]
@janpfeifer 
I think I didn't explain it clearly. Please take a look at this video(Sorry the video isn't very clear) and I think you will understand what I mean.

gdlv is the front end of Delve, just like ddd is the front end of gdb. We don't need jupyter when debugging, because you have already organized them all and put them in a main.go!

we just let gdlv debug the main.go

https://github.com/janpfeifer/gonb/assets/54462/7974ec1c-197c-481d-a122-7145d2997683

for CGO, maybe it is not fully featured(https://github.com/go-delve/delve/issues/935)

**Reply by janpfeifer** [reply]
Oh, thanks you for the explanation (and patience!)

I misunderstood the first message, I thought dlv didn't have a GUI, and it was a command line thing.

Also ... because I mostly use JupyterLab on a server and the browser in my laptop, GUI tools don't work as easily and it didn't occur to me -- although with `ssh -X` forwarding X11 protocol, it could work as well.

But this sounds a great use case, I'm going to start using it as well (at least when I don't need CGO!)

Now, I found that this line works as well:

```
!*gdlv debug "$(basename $(pwd))"
```

(notice `!*` executes the bash line in the temporary directory where the cell code is built and executed. Alternatively, one can use `GONB_TMP_DIR` that is set by gonb to that directory and is available)

Definitely a functionality we want to make easily and conveniently available.

Now the question is: is adding another special command in `gonb` the best solution to provide the functionality ?

As an alternative, one could include in their `.bashrc` something like:

```
alias dbg='cd ${GONB_TMP_DIR} ; gdlv debug "$(basename ${GONB_TMP_DIR})"'
```

(Or create a similarly named one line bash script file in the `PATH` somewhere).

And then in `gonb`:

```
!dbg
```

Wdyt ?

This is the way I thought folks would have their own custom/favourite aliases in `gonb` (as opposed to having to change `gonb`).

But maybe the integration of a Go debugger is so special that it is deserving of a special case ?

Or maybe there should be a way to easily configure aliases set only when executing the gonb kernel, that would be available inside cells ? (Maybe simply make gonb include `~/bin/gonb` to the path, if it exists ?)

**Reply by HaveF** [reply]
wow!

- `!*` is a nice solution! I didn't expect to use it this way
- `GONB_TMP_DIR` is also nice!

I think it's fine. Maybe we don't need another command. But I believe it is a useful tip for doc :-)

Thanks! @janpfeifer

**Reply by janpfeifer** [reply]
Cool, thanks for pointing this out to me @HaveF , this is already being handy here.

For `gonb` I added a small script [`ndlv`](https://github.com/janpfeifer/gonb/blob/jupytext/cmd/ndlv/ndlv) that checks the directories and that the programs are installed with helpful error messages. (*)

And I added a section in the `%help` message and in the tutorial. It will all show up in the next release when we are finished/happy with `jupytext` changes.

(*) I called it `ndlv` as a short for "Notebook delve" ... and to make it short/easy to type. I considered `gonb_gdlv`, but it seemed too long. If you have a better name suggestion let me know! :smiley:

**Reply by HaveF** [reply]
@janpfeifer ndlv is fine. That's great to have debug tools now. Except that I haven't learned how to use gonb to make packages yet, I'm very, very satisfied.

Maybe I'll experiment with it and see if there are any easy-to-implement ideas that we can borrow from nbdev (which I mentioned to you before). If I have some ideas, I'll come back to discuss it.

Thanks again!

**Reply by janpfeifer** [reply]
No problems, thank you for the good suggestion.

Any other ideas that can be handy let me know!

**Comment by janpfeifer** [comment]
Btw, v0.10.2 released with the `ndlv` script, and instructions in `%help` and in the tutorial.

Closing the topic for now!
